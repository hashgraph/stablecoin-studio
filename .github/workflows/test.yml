name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test