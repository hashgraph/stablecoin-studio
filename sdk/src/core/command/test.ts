/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
