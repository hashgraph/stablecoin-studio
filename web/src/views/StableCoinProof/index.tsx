import { VStack, Button, useDisclosure, Heading, Text, SimpleGrid, Flex } from '@chakra-ui/react';

import { UpdateReserveAddressRequest, UpdateReserveAmountRequest } from 'hedera-stable-coin-sdk';

import type { ReactNode } from 'react';
import { useState } from 'react';

import type { FieldValues } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import AwaitingWalletSignature from '../../components/AwaitingWalletSignature';
import BaseContainer from '../../components/BaseContainer';
import InputController from '../../components/Form/InputController';
import ModalNotification from '../../components/ModalNotification';
import { propertyNotFound } from '../../constant';

import { useRefreshCoinInfo } from '../../hooks/useRefreshCoinInfo';
import SDKService from '../../services/SDKService';
import {
	SELECTED_TOKEN_RESERVE_ADDRESS,
	SELECTED_TOKEN_RESERVE_AMOUNT,
	SELECTED_WALLET_COIN,
} from '../../store/slices/walletSlice';
import { handleRequestValidation } from '../../utils/validationsHelper';
import NoProofOfReserve from './components/NoProofOfReserve';

const StableCoinProof = () => {
	const [awaitingUpdate, setAwaitingUpdate] = useState<boolean>(false);
	const [success, setSuccess] = useState<boolean>();
	const [error, setError] = useState<any>();
	const { isOpen, onOpen, onClose } = useDisclosure();
	const variant = awaitingUpdate ? 'loading' : success ? 'success' : 'error';

	const { t } = useTranslation(['proofOfReserve', 'global']);

	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const reserveAddress = useSelector(SELECTED_TOKEN_RESERVE_ADDRESS);
	const reserveAmount = useSelector(SELECTED_TOKEN_RESERVE_AMOUNT);

	const [updateReserveAddressRequest] = useState<UpdateReserveAddressRequest>(
		new UpdateReserveAddressRequest({
			tokenId: '',
			reserveAddress: '',
		}),
	);
	const [updateReserveAmountRequest] = useState<UpdateReserveAmountRequest>(
		new UpdateReserveAmountRequest({
			reserveAddress: '',
			reserveAmount: '',
		}),
	);

	const isLoading = useRefreshCoinInfo();

	const form = useForm<FieldValues>({
		mode: 'onChange',
	});

	const { control, getValues } = form;

	const handleUpdateReserveAddress = async () => {
		const { reserveAddress } = getValues();
		if (selectedStableCoin?.tokenId) {
			updateReserveAddressRequest.tokenId = selectedStableCoin.tokenId.toString();
			updateReserveAddressRequest.reserveAddress = reserveAddress;

			try {
				onOpen();
				setAwaitingUpdate(true);
				await SDKService.updateReserveAddress(updateReserveAddressRequest);
				setError('');
				setAwaitingUpdate(false);
				setSuccess(true);
			} catch (error: any) {
				setAwaitingUpdate(false);
				console.log(error);
				setError(error?.transactionError?.transactionUrl);
				setSuccess(false);
				setAwaitingUpdate(false);
			}
		}
	};

	const handleUpdateReserveAmount = async () => {
		const { updateReserveAmount } = getValues();
		updateReserveAmountRequest.reserveAddress = reserveAddress?.toString() ?? '';
		updateReserveAmountRequest.reserveAmount = updateReserveAmount;

		try {
			onOpen();
			setAwaitingUpdate(true);
			await SDKService.updateReserveAmount(updateReserveAmountRequest);
			setError('');
			setAwaitingUpdate(false);
			setSuccess(true);
		} catch (error: any) {
			setAwaitingUpdate(false);
			console.log(error);
			setError(error?.transactionError?.transactionUrl);
			setSuccess(false);
			setAwaitingUpdate(false);
		}
	};

	const GridItem = ({
		name,
		title,
		label,
		current,
		input,
		button,
	}: {
		name: string;
		title: string;
		label: string;
		current: string;
		input: ReactNode;
		button: ReactNode;
	}) => (
		<Flex direction='column' alignItems={'center'}>
			<VStack w='100%' alignItems={'start'} gap='6px' maxWidth={'500px'} alignSelf='center'>
				<Heading
					data-testid={name + '-title'}
					fontSize='24px'
					fontWeight='600'
					mb={8}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{title}
				</Heading>
				<Text data-testid={name + '-label'} style={{ fontWeight: '600' }}>
					{label}
				</Text>
				<Text
					pl='10px'
					data-testid={name + '-value'}
					color={current === 'Error' ? 'red' : ''}
					mb={5}
				>
					{current}
				</Text>
				{input}
				{button}
			</VStack>
		</Flex>
	);

	return (
		<BaseContainer title={t('proofOfReserve:title')}>
			{isLoading && <AwaitingWalletSignature />}
			{selectedStableCoin && !isLoading && selectedStableCoin.reserveAddress && (
				<Flex
					direction='column'
					bg='brand.gray100'
					px={{ base: 4, lg: 14 }}
					pt={{ base: 4, lg: 14 }}
					pb={6}
				>
					<SimpleGrid columns={{ base: 2 }} gap={{ base: 10, lg: 20, md: 40 }} alignItems='center'>
						{GridItem({
							name: 'amount',
							title: t('proofOfReserve:updateReserveAmount.title'),
							label: t('proofOfReserve:updateReserveAmount.label'),
							current: reserveAmount?.toString() ?? '',
							input: (
								<InputController
									control={control}
									rules={{
										required: t('global:validations.required') ?? propertyNotFound,
										validate: {
											validation: (value: string) => {
												updateReserveAmountRequest.reserveAmount = value;
												const res = handleRequestValidation(
													updateReserveAmountRequest.validate('reserveAmount'),
												);
												return res;
											},
										},
									}}
									name={'updateReserveAmount'}
									placeholder={
										t('proofOfReserve:updateReserveAmount.inputPlaceholder') ?? propertyNotFound
									}
									isReadOnly={false}
								/>
							),
							button: (
								<Button
									data-testid={`update-reserve-amount-button`}
									variant='primary'
									onClick={handleUpdateReserveAmount}
								>
									{t('proofOfReserve:updateReserveAmount.buttonText')}
								</Button>
							),
						})}
						{GridItem({
							name: 'address',
							title: t('proofOfReserve:updateReserveAddress.title'),
							label: t('proofOfReserve:updateReserveAddress.label'),
							current: reserveAddress?.toString() ?? '',
							input: (
								<InputController
									rules={{
										required: t('global:validations.required') ?? propertyNotFound,
										validate: {
											validation: (value: string) => {
												updateReserveAddressRequest.reserveAddress = value;
												const res = handleRequestValidation(
													updateReserveAddressRequest.validate('reserveAddress'),
												);
												return res;
											},
										},
									}}
									control={control}
									name={'reserveAddress'}
									placeholder={
										t('proofOfReserve:updateReserveAddress.inputPlaceholder') ?? propertyNotFound
									}
								/>
							),
							button: (
								<Button
									data-testid={`update-reserve-address-button`}
									variant='primary'
									onClick={handleUpdateReserveAddress}
								>
									{t('proofOfReserve:updateReserveAddress.buttonText')}
								</Button>
							),
						})}
					</SimpleGrid>
				</Flex>
			)}
			<ModalNotification
				variant={variant}
				title={
					awaitingUpdate
						? 'Loading'
						: t('proofOfReserve:notification.title', {
								result: success ? 'Success' : 'Error',
						  })
				}
				description={
					awaitingUpdate
						? undefined
						: t(`proofOfReserve:notification.description${success ? 'Success' : 'Error'}`)
				}
				isOpen={isOpen}
				onClose={onClose}
				onClick={onClose}
				errorTransactionUrl={error}
				closeButton={false}
				closeOnOverlayClick={false}
			/>
			{selectedStableCoin && !isLoading && !selectedStableCoin.reserveAddress && (
				<NoProofOfReserve />
			)}
		</BaseContainer>
	);
};

export default StableCoinProof;
