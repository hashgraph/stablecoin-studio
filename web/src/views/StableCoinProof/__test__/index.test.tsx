import { render, sdkMock } from '../../../test/index';
import translations from '../../../translations/en/proofOfReserve.json';
import configureMockStore from 'redux-mock-store';
import StableCoinProof from '../';
import { screen } from '@testing-library/react';

const mockStore = configureMockStore();

describe(`<${StableCoinProof.name} />`, () => {

	afterEach(() => {
		jest.clearAllMocks();
	})

	test('should render correctly', () => {
		const component = render(<StableCoinProof />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should have a title', () => {
		const component = render(<StableCoinProof />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});

	test('should have buttons', async () => {
		const selectedStableCoin = {
			initialSupply: 0,
			tokenId: '0.0.48162226',
			totalSupply: 0,
			name: 'MIDAS',
			symbol: 'MD',
			decimals: 3,
			id: '0.0.48132286',
			maxSupply: '100000',
			treasuryId: '0.0.48160285',
			memo: 'Hedera Accelerator Stable Coin',
			adminKey: {
				key: 'key',
				type: 'ED25519',
			},
			freezeKey: {
				key: 'key',
				type: 'ED25519',
			},
			wipeKey: {
				id: '0.0.48160285',
			},
			supplyKey: {
				id: '0.0.48160285',
			},
			reserveAddress: '0.0.444',
			reserveAmount: '0',
		};

		const store = mockStore({
			wallet: {
				selectedStableCoin,
			},
		});
		
		const getAddressMock = sdkMock('getReserveAddress');
		const getAmountMock = sdkMock('getReserveAmount');

		getAddressMock.mockReturnValueOnce('0.0.444');
		getAmountMock.mockReturnValueOnce('0');

		render(<StableCoinProof />, store);
		const amountButton = await screen.findByTestId('update-reserve-amount-button');
		expect(amountButton).toBeInTheDocument();
		expect(amountButton).toBeEnabled();
		const addressButton = await screen.findByTestId('update-reserve-address-button');
		expect(addressButton).toBeInTheDocument();
		expect(addressButton).toBeEnabled();
	});
});
