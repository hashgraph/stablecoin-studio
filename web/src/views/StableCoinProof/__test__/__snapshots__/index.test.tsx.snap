// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StableCoinProof /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <h2
      class="chakra-heading css-bqsusc"
      data-testid="base-container-heading"
    >
      Proof of Reserve
    </h2>
    <div
      class="css-1b6lda7"
      data-testid="base-container-children"
    >
      <div
        class="css-1jib1v"
        data-testid="no-proof-of-reserve-container"
      >
        <div
          class="chakra-spinner css-n6xvdu"
          data-testid="awaiting-wallet-signature-logo"
        >
          <span
            class="css-8b45rq"
          >
            Loading...
          </span>
        </div>
        <p
          class="chakra-text css-e8j0ix"
          data-testid="no-proof-of-reserve-title"
        >
          Loading data
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
